test/apps/iodemo/io_demo.cc

```diff
@@ -832,132 +889,182 @@ class DemoClient : public P2pDemoCommon {
         return tv.tv_sec + (tv.tv_usec * 1e-6);
     }
 
-    bool run() {
-        std::vector<UcxConnection*> conn;
-        conn.resize(opts().servers.size());
-        for (size_t i = 0; i < conn.size(); i++) {
-            conn[i] = connect(opts().servers[i]);
-            if (!conn[i]) {
-                LOG << "Connect to server ["
-                    << opts().servers[i]
-                    << "] Failed!";
-                for (size_t j = 0; j < i; j++) {
-                    delete conn[j];
-                }
-                return false;
+    const std::string server_name(unsigned server_index) {
+        std::stringstream ss;
+        ss << "server [" << server_index << "] " << opts().servers[server_index];
+        return ss.str();
+    }
+
+    void connect_failed(unsigned server_index) {
+        server_info_t& server_info = _server_info[server_index];
+
+        ++server_info.retry_count;
+
+        UcxLog log(LOG_PREFIX);
+        log << "Connect to " << server_name(server_index) << " failed"
+            << " (retry " << server_info.retry_count;
+        if (opts().client_retries < std::numeric_limits<long>::max()) {
+            log << "/" << opts().client_retries;
+        }
+        log << ")";
+
+        if (server_info.retry_count >= opts().client_retries) {
+            /* If at least one server exceeded its retries, bail */
+            _status = CONN_RETRIES_EXCEEDED;
+        }
+    }
+
+    void connect_all() {
+        for (int server_index = 0; server_index < _server_info.size();
+             ++server_index) {
+            server_info_t& server_info = _server_info[server_index];
+            if ((server_info.conn != NULL) ||
+                (server_info.retry_count >= opts().client_retries)) {
+                /* server is already connected, or retry count exceeded */
+                continue;
             }
+
+            server_info.conn = connect(opts().servers[server_index]);
```

Comment:
let's  move address `opts().servers[server_index]` to `server_info` and pass to `connect`, so we can link conn <-> server_info

Reply:
but opts() is part of argument parsing which happens at earlier stage

Reply:
I mean just initialize `const char* ` from `options_t::servers`. It should be valid during client life time